File: ts/src/types/http_handler.ts
import { HttpHandlerFunc } from "./http_handler_func.js";
import { HttpRequest } from "./http_request.js";
import { HttpResponseWriter } from "./http_response_writer.js";
import { HttpServer } from "./http_server.js";
import { Result } from "@adviser/cement";

export interface HTTPHandlerParam {
  readonly HttpServer: HttpServer;
}

const textEncoder = new TextEncoder();
export class HTTPHandler {
  readonly _params: HTTPHandlerParam;
  readonly _handlerMap = new Map<string, HttpHandlerFunc>();

  constructor(hp: HTTPHandlerParam) {
    this._params = hp;
    this._params.HttpServer.SetHandler({
      ServeHTTP: async (w: HttpResponseWriter, r: HttpRequest): Promise<void> => {
        try {
          const fn = this._handlerMap.get(r.URL.Path);
          if (fn) {
            await fn(w, r);
            await w.End();
            return Promise.resolve();
          }
          w.WriteHeader(404);
          await w.Write(textEncoder.encode("Not found: " + r.URL.Path));
          await w.End();
          return Promise.resolve();
        } catch (err) {
          return Promise.reject(err);
        }
      },
    });
  }

  HttpServer(): HttpServer {
    return this._params.HttpServer;
  }

  Start(): Promise<void> {
    return this._params.HttpServer.ListenAndServe();
  }

  Stop(): Promise<void> {
    return this._params.HttpServer.Shutdown();
  }

  RegisterHandler(path: string, fn: HttpHandlerFunc): Result<() => void, Error> {
    if (this._handlerMap.has(path)) {
      return Result.Err(new Error(`Handler already registered, path: ${path}`));
    }
    this._handlerMap.set(path, fn);
    return Result.Ok(() => {
      this._handlerMap.delete(path);
    });
  }
}

File: go/transport/http_handler/http_handler.go
package transport_http_handler

import (
	"context"
	"fmt"
	"net/http"
	"sync"

	"github.com/mabels/runified/types"
	types_app "github.com/mabels/runified/types/app"
)

type HTTPHandlerParam struct {
	Ctx        context.Context
	Listen     types_app.AddrPort
	HttpServer types.HttpServer
}
type HTTPHandler struct {
	srv        types.HttpServer
	ctx        context.Context
	params     HTTPHandlerParam
	handlerMap map[string]http.HandlerFunc
	done       sync.WaitGroup
}

type goHttpServer struct {
	srv *http.Server
}

func (gs *goHttpServer) SetHandler(h http.Handler) {
	gs.srv.Handler = h
}

func (gs *goHttpServer) ListenAndServe() error {
	return gs.srv.ListenAndServe()
}

func (gs *goHttpServer) Shutdown(ctx context.Context) error {
	return gs.srv.Shutdown(ctx)
}

func newGoHttpServer(ap types_app.AddrPort) types.HttpServer {
	return &goHttpServer{
		srv: &http.Server{
			Addr: fmt.Sprintf("%s:%d", ap.Addr, ap.Port),
		},
	}
}

func NewHTTPHandler(hp HTTPHandlerParam) *HTTPHandler {
	if hp.Ctx == nil {
		hp.Ctx = context.Background()
	}
	ctx, _ := context.WithCancel(hp.Ctx)
	if hp.HttpServer == nil {
		hp.HttpServer = newGoHttpServer(hp.Listen)
	}
	ret := &HTTPHandler{
		ctx:        ctx,
		params:     hp,
		srv:        hp.HttpServer,
		handlerMap: make(map[string]http.HandlerFunc),
	}
	ret.srv.SetHandler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		fn, found := ret.handlerMap[r.URL.Path]
		if found {
			fn(w, r)
		}
		//_404Page(w,r)
	}))
	return ret
}

func (h *HTTPHandler) Start() error {
	h.done.Add(1)
	err := h.srv.ListenAndServe()
	h.done.Done()
	return err
}

func (h *HTTPHandler) Stop() error {
	err := h.srv.Shutdown(h.ctx)
	h.done.Wait()
	return err
}

func (h *HTTPHandler) RegisterHandler(path string, fn http.HandlerFunc) (func(), error) {
	_, found := h.handlerMap[path]
	if found {
		return nil, fmt.Errorf("Handler already registered, path: %s", path)
	}
	h.handlerMap[path] = fn
	return func() {
		delete(h.handlerMap, path)
	}, nil
}

// func StartServer(port uint16) {
// 	contextLogger := zerolog.New(os.Stdout).With().Uint16("port", port).Timestamp().Logger()
// 	contextLogger.Info().Msg("StartServer called")

// 	//myHelloHandler := constructHandler("Hello")
// 	http.HandleFunc("/", HelloHandler)

// 	http.HandleFunc("/hi", HiHandler)

// 	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", port), nil))
// }
